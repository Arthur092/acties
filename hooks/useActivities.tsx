import { collection, getDoc, onSnapshot, query } from "firebase/firestore";
import React, { useState, useContext, createContext, useEffect } from "react";
import { ActivityType, RecordType } from "../constants/SampleData";
import { db, getActivityTypesByUser, getRecordsByUser } from "../firebase";
import { useAuth } from "./useAuth";

interface ContextValue {
    activityTypes: ActivityTypesState;
    records: RecordsState;
    getActivityTypes: Function;
    getRecords: Function
}

type ActivityTypesState = {
    data: Array<ActivityType>;
    isLoading: Boolean;
}

export type RecordsState = {
    data: Array<RecordType>;
    isLoading: Boolean;
}

const ActivitiesContext = createContext<ContextValue>({
    activityTypes: {
        data: [],
        isLoading: true
    },
    records: {
        data: [],
        isLoading: true
    },
    getActivityTypes: () => null,
    getRecords: () => null,
});

interface Props {
    children: JSX.Element
}

export function ProvideActivities({ children }: Props) {
    const activities: ContextValue = useProvideActivities();

    return (
        <ActivitiesContext.Provider value={activities}>
            { children }
        </ActivitiesContext.Provider>
    )
}

export const useActivities = () => useContext(ActivitiesContext);

function useProvideActivities(): ContextValue {
    const { user } = useAuth();
    const [newRecord, setNewRecord] = useState<RecordType | null>(null);

    useEffect(() => {
        const queryListener = query(collection(db, "Record"));
        const unsubscribe = onSnapshot(queryListener, (querySnapshot) => {
            querySnapshot.docChanges().forEach(change => {
                if (change.type === "added") {
                    const { activityType, date, quantity, userId } = change.doc.data();
                    const newRecord = {
                        id: change.doc.id,
                        activity: activityType,
                        date,
                        quantity,
                        userId
                    }
                    getDoc(newRecord.activity).then(activity => {
                        setNewRecord({
                            ...newRecord,
                            activity: activity.data() as ActivityType
                        });
                    });
                }
            })
        });

        return () => unsubscribe();
    },[]);

    useEffect(() => {
        if(newRecord){
            setRecords({
                ...records,
                data: [
                    ...records.data,
                    newRecord
                ],
            })
        }
    },[newRecord])

    const [activityTypes, setActivityTypes] = useState<ActivityTypesState>({
        data: [],
        isLoading: true
    });
    const [records, setRecords] = useState<RecordsState>({
        data: [],
        isLoading: true
    });

    const getActivityTypes = () => {
        if (user) {
            const { uid } = user;
            return getActivityTypesByUser(uid).then(newActivityTypes => {
                setActivityTypes({
                    ...activityTypes,
                    data: newActivityTypes,
                    isLoading: false
                })
            });
        }
    }

    const getRecords = () => {
        if (user) {
            const { uid } = user;
            return getRecordsByUser(uid).then(newRecords => {
                setRecords({
                    ...records,
                    data: newRecords,
                    isLoading: false
                })
            });
        }
    }

    return {
        activityTypes,
        records,
        getActivityTypes,
        getRecords
    }
}


