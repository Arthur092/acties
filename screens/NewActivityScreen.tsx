import * as React from 'react';
import { useState } from 'react';
import { StyleSheet, View } from 'react-native';
import { Provider, Divider, List, Snackbar } from 'react-native-paper';
import { AddRecordDialog } from '../components/NewRecordDialog';
import { useThemeColor } from '../components/Themed';
import { ActivityType } from '../constants/SampleData';
import { useActivities } from '../hooks/useActivities';
import { theme as coreTheme} from '../core/theme'

export function NewActivityScreen(){
  const [isDialogVisile, setIsDialogVisile] = useState(false);
  const [currentActivity, setCurrentActivity] = useState<ActivityType | null>(null);
  const [snackBar, setSnackBar] = useState({visible: false, message: '', error: false});

  const showDialog = (value: boolean, activity: ActivityType | null = null) => {
    if(activity){
      setCurrentActivity(activity);
    }
    setIsDialogVisile(value);
  };
  const { activityTypes } = useActivities();

  return (
    <Provider>
      <List.Section>
        {activityTypes.data.map((activity, index) => (
          <View key={index}>
            <Divider />
            <List.Item
              title={activity.name}
              left={() => <List.Icon icon={activity.iconName} color={activity.iconColor}/>}
              right={() => <List.Icon icon="plus"/>}
              onPress={() => showDialog(true, activity)}
              style={styles(useThemeColor).list}
            />
          </View>
        ))}
        <Divider />
      </List.Section>

      <AddRecordDialog
        visible={isDialogVisile}
        showDialog={showDialog}
        currentActivity={currentActivity}
        setSnackBar={setSnackBar}
      />
      <Snackbar
        visible={snackBar.visible}
        onDismiss={() => setSnackBar({...snackBar, visible: false})}
        style={snackBar.error ? styles(useThemeColor).snackBarError : styles(useThemeColor).snackBarSuccess}
        >
          {snackBar.message}
      </Snackbar>
    </Provider>
  );
}

const styles = (theme: Function) => StyleSheet.create({
  list: {
    backgroundColor: theme({}, 'background'),
  },
  snackBarError: {
    backgroundColor: coreTheme.colors.error
  },
  snackBarSuccess: {
    backgroundColor: coreTheme.colors.success
  },
});
