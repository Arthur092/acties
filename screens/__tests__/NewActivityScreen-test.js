import React, { Profiler } from 'react';
import renderer from 'react-test-renderer';
import {
  render,
  fireEvent,
  waitFor,
  waitForElementToBeRemoved,
} from '@testing-library/react-native';
import { NewActivityScreen } from '../NewActivityScreen';
import { ActivitiesContext } from '../../hooks/useActivities';

jest.useFakeTimers();

describe('Load NewActivityScreen', () => {
  it('should renders correctly', () => {
    const tree = renderer.create(<NewActivityScreen />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('When an activity is clicked', () => {
  const testActivityContextData = {
    activityTypes: {
      data: [
        {
          id: 'testid',
          name: 'testType',
          isQuantity: 1,
          iconName: 'gas-station',
          iconColor: 'red',
          userId: 'testUserId',
        },
      ],
      isLoading: false,
    },
    records: {
      data: [],
      isLoading: false,
    },
    getActivityTypes: () => null,
    getRecords: () => null,
  };

  it('should show the dialog', async () => {
    const { getByTestId, debug } = render(
      <ActivitiesContext.Provider value={testActivityContextData}>
        <NewActivityScreen />
      </ActivitiesContext.Provider>
    );

    fireEvent.press(getByTestId('list-0'));
    const newRecordDialog = await waitFor(() =>
      getByTestId('new-record-dialog')
    );
    expect(newRecordDialog).toBeDefined();
  });

  it('should dissmiss the dialog', async () => {
    const { getByTestId, debug } = render(
      <ActivitiesContext.Provider value={testActivityContextData}>
        <NewActivityScreen />
      </ActivitiesContext.Provider>
    );

    fireEvent.press(getByTestId('list-0'));
    const newRecordDialog = await waitFor(() =>
      getByTestId('new-record-dialog')
    );

    fireEvent.press(getByTestId('dialog-cancel-button'));
    const newRecordDialogAfter = await waitForElementToBeRemoved(() =>
      getByTestId('new-record-dialog')
    );

    expect(newRecordDialogAfter).toBeDefined();
  });
});
