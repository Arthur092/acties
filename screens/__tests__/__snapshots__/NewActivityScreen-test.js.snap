// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Load NewActivityScreen should renders correctly 1`] = `
<View
  collapsable={false}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "marginVertical": 8,
        },
        undefined,
      ]
    }
    theme={
      Object {
        "animation": Object {
          "scale": 1,
        },
        "colors": Object {
          "accent": "#03dac4",
          "backdrop": "rgba(0, 0, 0, 0.5)",
          "background": "#f6f6f6",
          "disabled": "rgba(0, 0, 0, 0.26)",
          "error": "#B00020",
          "notification": "#f50057",
          "onSurface": "#000000",
          "placeholder": "rgba(0, 0, 0, 0.54)",
          "primary": "#6200ee",
          "surface": "#ffffff",
          "text": "#000000",
        },
        "dark": false,
        "fonts": Object {
          "light": Object {
            "fontFamily": "System",
            "fontWeight": "300",
          },
          "medium": Object {
            "fontFamily": "System",
            "fontWeight": "500",
          },
          "regular": Object {
            "fontFamily": "System",
            "fontWeight": "400",
          },
          "thin": Object {
            "fontFamily": "System",
            "fontWeight": "100",
          },
        },
        "roundness": 4,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "rgba(0, 0, 0, 0.12)",
            "height": 0.5,
          },
          undefined,
          undefined,
        ]
      }
    />
  </View>
  <View
    collapsable={false}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View />
  </View>
</View>
`;
